xdsl/dialects/x86/ops.py

```diff
@@ -300,59 +300,61 @@ class RR_MovOp(R_RR_Operation[GeneralRegisterType, GeneralRegisterType]):
     name = "x86.rr.mov"
 
 
-class M_R_Operation(Generic[R1InvT], IRDLOperation, X86Instruction, ABC):
+@irdl_op_definition
+class R_PushOp(IRDLOperation, X86Instruction):
     """
-    A base class for x86 operations that have one source register.
+    Decreases %rsp and places r1 at the new memory location pointed to by %rsp.
+    https://www.felixcloutier.com/x86/push
     """
 
+    name = "x86.r.push"
+
+    rsp_input = operand_def(GeneralRegisterType("rsp"))
     source = operand_def(R1InvT)
+    rsp_output = result_def(GeneralRegisterType("rsp"))
 
     def __init__(
         self,
+        rsp_input: Operation | SSAValue,
         source: Operation | SSAValue,
         *,
         comment: str | StringAttr | None = None,
+        rsp_output: GeneralRegisterType,
     ):
         if isinstance(comment, str):
             comment = StringAttr(comment)
 
         super().__init__(
-            operands=[source],
+            operands=[rsp_input, source],
             attributes={
                 "comment": comment,
             },
-            result_types=[],
+            result_types=[rsp_output],
         )
 
     def assembly_line_args(self) -> tuple[AssemblyInstructionArg | None, ...]:
         return (self.source,)
 
 
 @irdl_op_definition
-class R_PushOp(M_R_Operation[GeneralRegisterType]):
+class R_PopOp(IRDLOperation, X86Instruction, ABC):
```

Comment:
```suggestion
class R_PopOp(IRDLOperation, X86Instruction):
```

Reply:
The issue is that ABC is supposed to override `__new__` and not let users init it directly, but so does our IRDL infrastructure, so we never catch this... All the ops in this file that have an `@irdl_op_definition` annotation should not be ABC

Reply:
oh right ok, I'll see if any stragglers slipped through in next PR as well.